src/basic/fs-util.c

```diff
@@ -1106,7 +1106,7 @@ int xopenat(int dir_fd, const char *path, int flags, mode_t mode) {
 
         if (isempty(path)) {
                 assert(!FLAGS_SET(flags, O_CREAT|O_EXCL));
-                return fd_reopen(dir_fd, flags);
+                return fd_reopen(dir_fd, flags & ~O_NOFOLLOW);
```

Comment:
hmm, this is probably something fd_reopen() should handle internally

i wonder what actually happens if you have an O_PATH fd onto a symlink, and then reopen it via /proc/self/fd/… Do you end up on the symlink or on the target of the symlink?